aptos-move/aptos-vm-types/src/module_write_set.rs

```diff
@@ -98,8 +98,13 @@ impl ModuleWriteSet {
         module_storage: &'a impl ModuleStorage,
     ) -> impl Iterator<Item = PartialVMResult<WriteOpInfo>> {
         self.writes.iter_mut().map(move |(key, write)| {
+            // MODULE METERING SAFETY:
+            //   If there is a write to the module at key K, it means the module at K has been read,
```

Comment:
This logic should hold up even with exists and size APIs, I suppose, since upgradability would have to be checked, right?

Reply:
Yes! I did not see a clean way to check this invariant here, as otherwise I would do
```rust
if module_storage.runtime_environment().vm_config().enable_lazy_loading {
  traversal_context.check_...(...)?;
}
```
but I thought that passing traversal context is an overkill given that we do charge for all write ops and it is not going to change.


Ideally, we would have in AptosVM something like:
1. charge for old if exists
2. charge for new
3. get all metadata, new/upgrade (for init module), etc.
but this requires quite a bit of changes throughout the system so kept like this for now.